File: packages/openmemory-js/src/temporal_graph/query.ts
/**
 * Temporal Knowledge Graph Query Engine for OpenMemory.
 * Provides APIs for point-in-time retrieval, range queries, and relationship traversal.
 */
import { env } from "../core/cfg";
import { q } from "../core/db";
import { getEncryption } from "../core/security";
import { normalizeUserId } from "../utils";
import { SimpleCache } from "../utils/cache";
import {
    TemporalEdge,
    TemporalEdgeRow,
    TemporalFact,
    TemporalFactRow,
} from "./types";

// Caches to avoid redundant decryption/parsing
export const factCache = new SimpleCache<string, TemporalFact>({
    maxSize: env.graphCacheSize,
    ttlMs: 1800000,
}); // 30 min
export const edgeCache = new SimpleCache<string, TemporalEdge>({
    maxSize: env.graphCacheSize * 2,
    ttlMs: 1800000,
});

/**
 * Converts a database row to a TemporalFact object.
 * Handles decryption of metadata and caching.
 * @param row The raw database row.
 * @returns The hydrated TemporalFact object.
 */
export async function rowToFact(
    row: TemporalFactRow | null,
): Promise<TemporalFact> {
    if (!row) return null as unknown as TemporalFact;

    // Cache key includes ID and last_updated to handle updates
    const cacheKey = `${row.id}:${row.lastUpdated}`;
    const cached = factCache.get(cacheKey);
    if (cached) return cached;

    const enc = getEncryption();
    let meta: Record<string, unknown> = {};
    if (row.metadata) {
        // Optimization: Check if it's already an object (Bun SQLite driver might parse JSON columns automatically)
        if (typeof row.metadata === 'object' && row.metadata !== null) {
            meta = row.metadata as Record<string, unknown>;
        } else if (typeof row.metadata === 'string') {
            try {
                // Priority 1: Decrypt if it looks encrypted (starts with "enc:")
                // Or just try decrypting.
                const asStr = row.metadata as string;
                if (asStr.startsWith('{"iv":')) {
                    const dec = await enc.decrypt(asStr);
                    meta = JSON.parse(dec);
                } else {
                    meta = JSON.parse(asStr);
                }
            } catch {
                // Fallback
                meta = { _raw: row.metadata };
            }
        }
    }
    const fact: TemporalFact = {
        id: row.id,
        userId: row.userId,
        subject: row.subject,
        predicate: row.predicate,
        object: row.object,
        validFrom: row.validFrom,
        validTo: row.validTo,
        confidence: row.confidence,
        lastUpdated: row.lastUpdated,
        metadata: meta,
    };

    factCache.set(cacheKey, fact);
    return fact;
}

/**
 * Converts a database row to a TemporalEdge object.
 * Handles decryption of metadata and caching.
 * @param row The raw database row.
 * @returns The hydrated TemporalEdge object.
 */
export async function rowToEdge(
    row: TemporalEdgeRow | null,
): Promise<TemporalEdge> {
    if (!row) return null as unknown as TemporalEdge;

    const cacheKey = `${row.id}:${row.lastUpdated}`;
    const cached = edgeCache.get(cacheKey);
    if (cached) return cached;

    const enc = getEncryption();
    let meta: Record<string, unknown> = {};
    if (row.metadata) {
        try {
            const dec = await enc.decrypt(row.metadata);
            meta = JSON.parse(dec);
        } catch {
            try {
                meta =
                    typeof row.metadata === "string"
                        ? JSON.parse(row.metadata)
                        : row.metadata;
            } catch {
                meta = { _raw: row.metadata };
            }
        }
    }
    const edge: TemporalEdge = {
        id: row.id,
        userId: row.userId,
        sourceId: row.sourceId,
        targetId: row.targetId,
        relationType: row.relationType,
        validFrom: row.validFrom,
        validTo: row.validTo,
        weight: row.weight,
        lastUpdated: row.lastUpdated,
        metadata: meta,
    };

    edgeCache.set(cacheKey, edge);
    return edge;
}

/**
 * Query facts valid at a specific point in time.
 * @param subject Optional subject filter
 * @param predicate Optional predicate filter
 * @param object Optional object filter
 * @param at Point in time to query (defaults to now)
 * @param minConfidence Minimum confidence threshold (default 0.0)
 * @param userId Owner of the facts (if null, checks for global facts; if undefined, checks all)
 * @returns Array of matching TemporalFact objects
 */
export const queryFactsAtTime = async (
    subject?: string,
    predicate?: string,
    object?: string,
    at: Date = new Date(),
    minConfidence: number = 0.0,
    userId?: string | null,
): Promise<TemporalFact[]> => {
    const timestamp = at.getTime();
    const uid = normalizeUserId(userId);

    const rows = await q.queryFactsAtTime.all(timestamp, subject, predicate, object, minConfidence, uid) as TemporalFactRow[];
    return Promise.all(rows.map(rowToFact));
};

/**
 * Get the single most current fact matching the criteria.
 * @param subject Exact subject
 * @param predicate Exact predicate
 * @param userId Owner ID
 * @param at Point in time (optional)
 * @returns The matching Fact or null
 */
export const getCurrentFact = async (
    subject: string,
    predicate: string,
    userId?: string | null,
    at?: Date,
): Promise<TemporalFact | null> => {
    const uid = normalizeUserId(userId);
    const timestamp = at?.getTime();

    const row = await q.getCurrentFact.get(subject, predicate, timestamp, uid) as TemporalFactRow | undefined;
    return await rowToFact(row || null);
};

/**
 * Query facts that were valid at any point within a time range.
 * @param subject Subject filter
 * @param predicate Predicate filter
 * @param object Object filter
 * @param from Start of range
 * @param to End of range
 * @param minConfidence Minimum confidence
 * @param userId User ID
 * @param limit Max results
 */
export const queryFactsInRange = async (
    subject?: string,
    predicate?: string,
    object?: string,
    from?: Date,
    to?: Date,
    minConfidence: number = 0.1,
    userId?: string | null,
    limit: number = 500,
): Promise<TemporalFact[]> => {
    const uid = normalizeUserId(userId);
    const fromTs = from?.getTime();
    const toTs = to?.getTime();

    const rows = await q.queryFactsInRange.all(fromTs, toTs, subject, predicate, object, minConfidence, limit, uid) as TemporalFactRow[];
    return Promise.all(rows.map(rowToFact));
};

/**
 * Find conflicting facts (same subject/predicate) valid at the same time.
 * Useful for consistency checks.
 * @param subject Subject
 * @param predicate Predicate
 * @param at Time to check
 * @param userId User ID
 */
export const findConflictingFacts = async (
    subject: string,
    predicate: string,
    at: Date = new Date(),
    userId?: string | null,
): Promise<TemporalFact[]> => {
    const timestamp = at ? at.getTime() : Date.now();
    const uid = normalizeUserId(userId);

    const rows = await q.findConflictingFacts.all(subject, predicate, timestamp, uid) as TemporalFactRow[];
    return Promise.all(rows.map(rowToFact));
};

/**
 * Get all facts known about a subject.
 * @param subject The entity subject
 * @param at Time (if not historical)
 * @param includeHistorical If true, returns all history regardless of validity
 * @param userId User ID
 * @param limit Limit results
 */
export const getFactsBySubject = async (
    subject: string,
    at: Date = new Date(),
    includeHistorical: boolean = false,
    userId?: string | null,
    limit: number = 500,
): Promise<TemporalFact[]> => {
    const uid = normalizeUserId(userId);
    const timestamp = at ? at.getTime() : Date.now();

    const rows = await q.getFactsBySubject.all(subject, timestamp, includeHistorical, limit, uid) as TemporalFactRow[];
    return Promise.all(rows.map(rowToFact));
};

/**
 * Search facts using SQL LIKE matching.
 * @param pattern The search pattern (do not include wildcards, they are added automatically)
 * @param type Search scope: "subject", "predicate", "object", or "all"
 * @param at Time constraint
 * @param limit Max results
 * @param userId User ID
 */
export const searchFacts = async (
    pattern: string,
    type: "subject" | "predicate" | "object" | "all" = "all",
    at?: Date,
    limit: number = 10,
    userId?: string | null,
): Promise<TemporalFact[]> => {
    const timestamp = at?.getTime();
    const uid = normalizeUserId(userId);

    const rows = await q.searchFacts.all(pattern, type, timestamp, limit, uid) as TemporalFactRow[];
    return Promise.all(rows.map(rowToFact));
};

/**
 * Get facts directly connected to a specific fact via Edges.
 * Represents a 1-hop neighborhood in the knowledge graph.
 * @param factId The source fact ID
 * @param relationType Optional edge type filter
 * @param at Time constraint
 * @param userId User ID
 */
export const getRelatedFacts = async (
    factId: string,
    relationType?: string,
    at?: Date,
    userId?: string,
): Promise<Array<{ fact: TemporalFact; relation: string; weight: number }>> => {
    const timestamp = at ? at.getTime() : Date.now();
    const uid = normalizeUserId(userId);

    // Repo returns row plus relation_type and weight
    const rows = await q.getRelatedFacts.all(factId, relationType, timestamp, uid) as (TemporalFactRow & { relation_type: string; weight: number })[];

    return Promise.all(
        rows.map(async (row) => ({
            fact: await rowToFact(row),
            relation: row.relation_type,
            weight: row.weight,
        })),
    );
};

/**
 * Query edges (relationships between facts) with filters.
 * @param sourceId Filter by source fact ID
 * @param targetId Filter by target fact ID
 * @param relationType Filter by relationship type
 * @param at Time constraint
 * @param userId User ID
 * @param limit Limit
 * @param offset Offset
 */
export const queryEdges = async (
    sourceId?: string,
    targetId?: string,
    relationType?: string,
    at?: Date,
    userId?: string | null,
    limit: number = 100,
    offset: number = 0,
): Promise<TemporalEdge[]> => {
    const timestamp = at ? at.getTime() : Date.now();
    const uid = normalizeUserId(userId);

    const rows = await q.queryEdges.all(sourceId, targetId, relationType, timestamp, limit, offset, uid) as TemporalEdgeRow[];
    return Promise.all(rows.map(rowToEdge));
};

/**
 * Get all facts that were valid during the specified window.
 * @param startTime Start of window
 * @param endTime End of window
 * @param userId User ID
 */
export const getFactsInRange = async (
    startTime: Date,
    endTime: Date,
    userId?: string | null,
): Promise<TemporalFact[]> => {
    // Reusing queryFactsInRange which supports range
    return await queryFactsInRange(undefined, undefined, undefined, startTime, endTime, 0, userId, 10000);
};

File: packages/openmemory-py/src/openmemory/temporal_graph/query.py
import time
import os
import asyncio

def enforce_tenant(user_id: str = None) -> str:
    if user_id:
        return user_id
    if os.environ.get("OM_ALLOW_ANONYMOUS_TENANT") == "true":
        return "anonymous"
    raise ValueError("MissingTenantError: user_id is required for multi-tenant isolation.")

import json
from typing import List, Dict, Any, Optional

from ..core.db import db

async def query_facts_at_time(subject: Optional[str] = None, predicate: Optional[str] = None, subject_object: Optional[str] = None, at: int = None, min_confidence: float = 0.1, user_id: Optional[str] = None) -> List[Dict[str, Any]]:
    user_id = enforce_tenant(user_id)
    ts = at if at is not None else int(time.time()*1000)
    conds = ["user_id = ?", "(valid_from <= ? AND (valid_to IS NULL OR valid_to >= ?))"]
    params = [user_id, ts, ts]
    if subject:
        conds.append("subject = ?")
        params.append(subject)
    if predicate:
        conds.append("predicate = ?")
        params.append(predicate)
    if subject_object:
        conds.append("object = ?")
        params.append(subject_object)
    if min_confidence > 0:
        conds.append("confidence >= ?")
        params.append(min_confidence)

    sql = f"""
        SELECT id, user_id, subject, predicate, object, valid_from, valid_to, confidence, last_updated, metadata
        FROM temporal_facts
        WHERE {' AND '.join(conds)}
        ORDER BY confidence DESC, valid_from DESC
    """
    rows = await asyncio.to_thread(db.fetchall, sql, tuple(params))
    return [format_fact(r) for r in rows]

async def get_current_fact(subject: str, predicate: str, user_id: str = None) -> Optional[Dict[str, Any]]:
    user_id = enforce_tenant(user_id)
    sql = """
        SELECT id, subject, predicate, object, valid_from, valid_to, confidence, last_updated, metadata
        FROM temporal_facts
        WHERE subject = ? AND predicate = ? AND user_id = ? AND valid_to IS NULL
        ORDER BY valid_from DESC
        LIMIT 1
    """
    row = await asyncio.to_thread(db.fetchone, sql, (subject, predicate, user_id))
    if not row:
        return None
    return format_fact(row)

async def query_facts_in_range(subject: str = None, predicate: str = None, start: int = None, end: int = None, min_confidence: float = 0.1, user_id: str = None) -> List[Dict[str, Any]]:
    user_id = enforce_tenant(user_id)
    conds = ["user_id = ?"]
    params = [user_id]

    if start is not None and end is not None:
        conds.append("((valid_from <= ? AND (valid_to IS NULL OR valid_to >= ?)) OR (valid_from >= ? AND valid_from <= ?))")
        params.extend([end, start, start, end])
    elif start is not None:
        conds.append("valid_from >= ?")
        params.append(start)
    elif end is not None:
        conds.append("valid_from <= ?")
        params.append(end)

    if subject:
        conds.append("subject = ?")
        params.append(subject)
    if predicate:
        conds.append("predicate = ?")
        params.append(predicate)
    if min_confidence > 0:
        conds.append("confidence >= ?")
        params.append(min_confidence)

    where = f"WHERE {' AND '.join(conds)}" if conds else ""
    sql = f"""
        SELECT id, subject, predicate, object, valid_from, valid_to, confidence, last_updated, metadata
        FROM temporal_facts
        {where}
        ORDER BY valid_from DESC
    """
    rows = await asyncio.to_thread(db.fetchall, sql, tuple(params))
    return [format_fact(r) for r in rows]

async def find_conflicting_facts(subject: str, predicate: str, at: int = None, user_id: str = None) -> List[Dict[str, Any]]:
    user_id = enforce_tenant(user_id)
    ts = at if at is not None else int(time.time()*1000)
    sql = """
        SELECT id, subject, predicate, object, valid_from, valid_to, confidence, last_updated, metadata
        FROM temporal_facts
        WHERE subject = ? AND user_id = ? AND predicate = ? AND user_id = ?
        AND (valid_from <= ? AND (valid_to IS NULL OR valid_to >= ?))
        ORDER BY confidence DESC
    """
    rows = await asyncio.to_thread(db.fetchall, sql, (subject, predicate, user_id, ts, ts))
    return [format_fact(r) for r in rows]

async def get_facts_by_subject(subject: str, at: int = None, include_historical: bool = False, user_id: str = None) -> List[Dict[str, Any]]:
    user_id = enforce_tenant(user_id)
    params = [subject, user_id]
    if include_historical:
        sql = """
            SELECT id, subject, predicate, object, valid_from, valid_to, confidence, last_updated, metadata
            FROM temporal_facts
            WHERE subject = ? AND user_id = ?
            ORDER BY predicate ASC, valid_from DESC
        """
    else:
        ts = at if at is not None else int(time.time()*1000)
        sql = """
            SELECT id, subject, predicate, object, valid_from, valid_to, confidence, last_updated, metadata
            FROM temporal_facts
            WHERE subject = ? AND user_id = ?
            AND (valid_from <= ? AND (valid_to IS NULL OR valid_to >= ?))
            ORDER BY predicate ASC, confidence DESC
        """
        params.extend([ts, ts])

    rows = await asyncio.to_thread(db.fetchall, sql, tuple(params))
    return [format_fact(r) for r in rows]

async def search_facts(pattern: str, field: str = "subject", at: int = None, user_id: str = None) -> List[Dict[str, Any]]:
    user_id = enforce_tenant(user_id)
    ts = at if at is not None else int(time.time()*1000)
    search_pat = f"%{pattern}%"
    if field not in ["subject", "predicate", "object"]: field = "subject"

    sql = f"""
        SELECT id, subject, predicate, object, valid_from, valid_to, confidence, last_updated, metadata
        FROM temporal_facts
        WHERE {field} LIKE ? AND user_id = ?
        AND (valid_from <= ? AND (valid_to IS NULL OR valid_to >= ?))
        ORDER BY confidence DESC, valid_from DESC
        LIMIT 100
    """
    rows = await asyncio.to_thread(db.fetchall, sql, (search_pat, user_id, ts, ts))
    return [format_fact(r) for r in rows]

async def get_related_facts(fact_id: str, relation_type: str = None, at: int = None, user_id: str = None) -> List[Dict[str, Any]]:
    user_id = enforce_tenant(user_id)
    ts = at if at is not None else int(time.time()*1000)
    conds = ["e.user_id = ?", "f.user_id = ?", "(e.valid_from <= ? AND (e.valid_to IS NULL OR e.valid_to >= ?))"]
    params = [user_id, user_id, ts, ts]

    if relation_type:
        conds.append("e.relation_type = ?")
        params.append(relation_type)

    sql = f"""
        SELECT f.*, e.relation_type, e.weight
        FROM temporal_edges e
        JOIN temporal_facts f ON e.target_id = f.id
        WHERE e.source_id = ?
        AND {' AND '.join(conds)}
        AND (f.valid_from <= ? AND (f.valid_to IS NULL OR f.valid_to >= ?))
        ORDER BY e.weight DESC, f.confidence DESC
    """
    params.insert(0, fact_id)
    params.extend([ts, ts])

    rows = await asyncio.to_thread(db.fetchall, sql, tuple(params))
    return [{
        "fact": format_fact(r),
        "relation": r["relation_type"],
        "weight": r["weight"]
    } for r in rows]

def format_fact(row: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "id": row["id"],
        "user_id": row.get("user_id"),
        "subject": row["subject"],
        "predicate": row["predicate"],
        "object": row["object"],
        "valid_from": row["valid_from"],
        "valid_to": row["valid_to"],
        "confidence": row["confidence"],
        "last_updated": row["last_updated"],
        "metadata": json.loads(row["metadata"]) if row["metadata"] else None
    }
